Added net472 target

mafurman opened the pull request:
Addresses #1507 

The net472 ECDsaAdaper calls CreateECDsaUsingECParams (similar to the netstandard2_0 target).

However, the AsymmetricAdapter still makes use of RsaCryptoProviderProxy on the net472 target.

mafurman committed:
Added net472 target

GeoK commented on test/Microsoft.IdentityModel.Tokens.Tests/SecurityKeyTests.cs:
```diff
@@ -198,7 +198,7 @@ public static TheoryData<SecurityKeyTheoryData> CreateInternalIdsTestCases
                         TestId = nameof(KeyingMaterial.JsonWebKeyRsa_2048_Public)
                     },
-#if NET_CORE
+#if NET472 || NET_CORE
```
nit: comment on ln216 is no longer true.

mafurman replied:
Good catch. Fixed!

GeoK commented on test/Microsoft.IdentityModel.Tokens.Tests/CryptoProviderCacheTests.cs:
```diff
@@ -271,7 +271,7 @@ public static TheoryData<CryptoProviderCacheTheoryData> TryAddTheoryData
                         TestId = nameof(KeyingMaterial.DefaultSymmetricSecurityKey_256)
                     },
-#if NET_CORE
+#if NET472 || NET_CORE 
```
nit: comment on ln275 is no longer true

mafurman replied:
Fixed!

GeoK commented:
Are there any changes required here?
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/30c16ab82cf4338c6d6e04a5d4094795aa83c587/src/Microsoft.IdentityModel.Xml/Microsoft.IdentityModel.Xml.csproj#L33

GeoK commented:
[Pss ](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.rsasignaturepadding.pss?view=net-5.0)is available on net472
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/30c16ab82cf4338c6d6e04a5d4094795aa83c587/test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs#L1507

GeoK commented:
We should test RsaCryptoServiceProviderProxy using net472, imo. RSA keys could be created of X509 certs, where keys are created with CAPI and old cryptographic providers that don't know about sha2.

https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/30c16ab82cf4338c6d6e04a5d4094795aa83c587/test/Microsoft.IdentityModel.Tokens.Tests/RsaCryptoServiceProviderProxyTests.cs#L28

GeoK commented on test/Microsoft.IdentityModel.Tokens.Tests/ECDsaSecurityKeyTests.cs:
```diff
@@ -74,7 +74,7 @@ public void Defaults()
         public void CanComputeJwkThumbprint()
         {
-#if NET_CORE
+#if NET472 || NET_CORE
```
nit: messages below are no longer true.

GeoK commented on src/Microsoft.IdentityModel.Tokens/AsymmetricAdapter.cs:
```diff
@@ -170,7 +170,7 @@ internal byte[] Decrypt(byte[] data)
 
             // NET461 could have been passed RSACryptoServiceProvider
-#if NET461
+#if NET461 || NET472
```
nit: comment above could be updated. same with ln 231.

mafurman replied:
Fixed!

GeoK commented:
We should update these to reflect the changes:
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/Supported-Algorithms
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/IDX10690

GeoK commented on src/Microsoft.IdentityModel.Tokens/ECDsaSecurityKey.cs:
```diff
@@ -116,7 +116,7 @@ public override int KeySize
         public override bool CanComputeJwkThumbprint()
         {
-#if NETSTANDARD2_0
+#if NET472 || NETSTANDARD2_0
```
net472 supports [ECParameters](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.ecparameters?view=net-5.0#moniker-applies-to), there is no need for this check. Similar applies to ln 133.

You can also keep this logic as is, but refactor SupportsECParameters to return **true** on net472, without the unnecessary check.

mafurman replied:
I've refactored SupportsECParameters to return true.

mafurman commented:
> 
> 
> Are there any changes required here?
> 
> https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/30c16ab82cf4338c6d6e04a5d4094795aa83c587/src/Microsoft.IdentityModel.Xml/Microsoft.IdentityModel.Xml.csproj#L33

I don't think so? It doesn't seem like those two assemblies are directly referenced in this project.

> 
> 
> [Pss ](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.rsasignaturepadding.pss?view=net-5.0)is available on net472
> 
> https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/30c16ab82cf4338c6d6e04a5d4094795aa83c587/test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs#L1507

Fixed!




> 
> 
> We should test RsaCryptoServiceProviderProxy using net472, imo. RSA keys could be created of X509 certs, where keys are created with CAPI and old cryptographic providers that don't know about sha2.
> 
> https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/30c16ab82cf4338c6d6e04a5d4094795aa83c587/test/Microsoft.IdentityModel.Tokens.Tests/RsaCryptoServiceProviderProxyTests.cs#L28

Good point, thanks! :) I think this was left over from before we realized RsaCryptoServiceProviderProxy was necessary on the net472 target.



> 
> 
> We should update these to reflect the changes:
> https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/Supported-Algorithms
> https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/IDX10690

I created a separate issue to track this so I can add this in after the PR is merged in. :)
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1556

GeoK reviewed (approved):
LGTM. Still unsure about Microsoft.IdentityModel.Xml.csproj; should we add references (Microsoft.CSharp and System.Net.Http) to 472 or remove them for net45 and net461?

brentschmaltz commented on test/Microsoft.IdentityModel.Tokens.Tests/AsymmetricAdapterTests.cs:
```diff
@@ -39,7 +39,7 @@ namespace Microsoft.IdentityModel.Tokens.Tests
     public class AsymmetricAdapterTests
     {
         [Theory, MemberData(nameof(AsymmetricAdapterUsageTestCases))]
-        public void AsymmetricAdapterUsageTests(AsymmetricAdatperTheoryData theoryData)
```
thanks!

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/ECDsaAdapter.cs:
```diff
@@ -52,12 +52,14 @@ internal class ECDsaAdapter
-#if NETSTANDARD2_0
+#if NET472
+            CreateECDsaFunction = CreateECDsaUsingECParams;
+#elif NETSTANDARD2_0
```
@mafurman Could you add a comment here indicating that although NETSTANDARD2_0 specifies that ECParameters is supported, we still need to call SupportsECParameters() as 4..6.1 is listed as supporting NETSTANDARD2_0, but DOES NOT support ECParameters. see: https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.ecparameters?view=netstandard-2.0

This may help future devs as they may not understand the issues.

brentschmaltz reviewed (approved):
suggested adding a comment.

mafurman commented:
@GeoK I've gone ahead and removed the Microsoft.CSharp and System.Net.Http references from the net45 and net461 targets in Microsoft.IdentityModel.Xml.csproj. I think this is something we may want to do for the other .csproj files we have as well, so I created a separate issue to track this: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1572

@brentschmaltz That's a good suggestion. I added your comment in.

mafurman commented:
/azp run
